src/engine/routing_algorithms/map_matching.cpp

```diff
@@ -248,8 +263,48 @@ SubMatchingList mapMatching(SearchEngineData<Algorithm> &engine_working_data,
                         current_lengths[s_prime] = network_distance;
                         current_pruned[s_prime] = false;
                         model.breakage[t] = false;
+                        current_segment_id[s_prime] = last_id;
                     }
                 }
+                // Here we can add previous states to the current viterbi if we suppose, that user
+                // can stay
+                double precision = DEFAULT_GPS_PRECISION;
+                if (!trace_gps_precision.empty())
+                    precision = (*trace_gps_precision[t]);
+                const auto distance = util::coordinate_calculation::haversineDistance(
+                    candidates_list[prev_unbroken_timestamp][s].phantom_node.location,
+                    trace_coordinates[t]);
+
+                // We can use a previous state as one of possible states if it is inside the
+                // precision range, except for the last state in trace
+                if (distance < 3 * precision && t < candidates_list.size() - 1)
```

Comment:
I think we should try to move all of this logic out of the actual Viterbi loop. There is a high chance that the vectors will be modified multiple times per timestamp. That means a reallocation of `std::vector` is triggered and a lot of data is being copied. I suspect this is the major source of the slowdown you see.

To fix this, we can move the logic to the `src/engine/plugins/match.cpp` code that sets up the candidates. Actually this code is working against [this code](https://github.com/Project-OSRM/osrm-backend/blob/master/src/engine/plugins/match.cpp#L58-L71) which filters out these "stay" states.

So an equivalent implementation would be to remove that filter step, since the candidates that the R-Tree returns are already according to your criteria (`distance < 3 * precision`). This should be as fast or faster then the current implementation. There might be a few superfluous calls to `getNetworkDistance()`, but I think we could add a special case check to [that function](https://github.com/Project-OSRM/osrm-backend/blob/master/src/engine/routing_algorithms/routing_base_ch.cpp#L160) for `start == target` that skips some of the overhead of inserting the nodes in the heaps and calculates the distance only use the `PhantomNode`.